Source/SessionManager.swift

```diff
@@ -612,28 +660,38 @@ open class SessionManager {
                         )
                         encodingCompletion?(encodingResult)
                     }
-                } catch {
-                    DispatchQueue.main.async { encodingCompletion?(.failure(error)) }
                 }
+            } catch {
+                DispatchQueue.main.async { encodingCompletion?(.failure(error)) }
             }
         }
     }
 
     // MARK: Private - Upload Implementation
 
     private func upload(_ uploadable: UploadRequest.Uploadable) -> UploadRequest {
-        let task = uploadable.task(session: session, adapter: adapter, queue: queue)
-        let request = UploadRequest(session: session, task: task, originalTask: uploadable)
+        do {
+            let task = try uploadable.task(session: session, adapter: adapter, queue: queue)
+            let upload = UploadRequest(session: session, requestTask: .upload(uploadable, task))
 
-        if case let .stream(inputStream, _) = uploadable {
-            request.delegate.taskNeedNewBodyStream = { _, _ in inputStream }
-        }
+            if case let .stream(inputStream, _) = uploadable {
+                upload.delegate.taskNeedNewBodyStream = { _, _ in inputStream }
+            }
 
-        delegate[request.delegate.task] = request
+            delegate[task] = upload
 
-        if startRequestsImmediately { request.resume() }
+            if startRequestsImmediately { upload.resume() }
```

Comment:
Similar to my suggestion for `request`, you can probably stick this in a `defer`.